src/Map.elm

```diff
@@ -138,92 +143,134 @@ type Action
+                    ]
+              )
+            , ( LandscapeName "Hill"
+              , LandscapeResources
+                    [ LandscapeResource ( Mine, rockResources ) ]
+              )
+            ]
+        )
 
 
-generateLocationData : Int -> LandscapeId -> Maybe LocationData
-generateLocationData seed (LandscapeId landscapeId) =
+generateLocationData : LandscapeId -> Generator (Maybe LocationData)
+generateLocationData (LandscapeId landscapeId) =
     let
-        ( _, landscapeValue ) =
-            findSafeInDict landscapeId landscapes
-
-        ( randomLandscapeResourceTypeId, resourceTypeSeed ) =
-            step (Random.int 0 (List.length landscapeValue - 1)) (initialSeed seed)
+        (Landscapes l) =
+            landscapes
 
-        ( landscapeAction, landscapeResources ) =
-            findSafeInList randomLandscapeResourceTypeId landscapeValue
-
-        ( resourceId, resourceSeed ) =
-            step (Random.int 0 (Array.length landscapeResources - 1)) resourceTypeSeed
+        ( _, landscapeResources ) =
+            findSafeInDict landscapeId l
+    in
+    genRandomLandscapeResource landscapeResources
+        |> Random.andThen
+            (\(LandscapeResource ( resourceAction, resources )) ->
+                genRandomResource resources
+                    |> Random.andThen
+                        (\(Resource ( ResourceName resourceName, Rarity rarity, maxAmount )) ->
+                            genAmountChance maxAmount
+                                |> Random.andThen
+                                    (\( amount, luck ) ->
+                                        if amount > 0 && succeed luck rarity then
+                                            let
+                                                r =
+                                                    ( Resource ( ResourceName resourceName, Rarity rarity, maxAmount )
+                                                    , resourceAction
+                                                    )
+                                            in
+                                            Random.constant (Just ( r, Amount amount ))
+
+                                        else
+                                            Random.constant
+                                                (Debug.log
+                                                    (stringifyLandscapeFailure resourceName amount luck)
+                                                    Nothing
+                                                )
+                                    )
+                        )
+            )
 
-        resource : Resource
-        resource =
-            findSafeInDict resourceId landscapeResources
 
-        ( resourceName, Rarity rarity, Amount maxAmount ) =
-            resource
+genRandomLandscapeResource : LandscapeResources -> Generator LandscapeResource
+genRandomLandscapeResource (LandscapeResources landscapeResources) =
+    Random.andThen
```

Comment:
If you use Random.map instead of Random.andThen here you won't have to wrap the result of the expression in a Random.constant

Reply:
Thanks